Add GDScript warning pages (#10635)
godotengine/godot-docs: tutorials/scripting/gdscript/warnings/get_node_default_without_onready.rst

GET_NODE_DEFAULT_WITHOUT_ONREADY
================================

> This warning appears when a node's property is initialized to a location in the scene tree without using the ``@onready`` annotation.

Reviewer: ```suggestion
This warning appears when a node's property is initialized to a node in the scene tree without using the ``@onready`` annotation.
```
I'd say

Depending on how you attempt to access the scene tree, the warning message will be one of the following:

.. code-block:: none

    The default value uses "get_node()" which won't return nodes in the scene tree before "_ready()" is called. Use the "@onready" annotation to solve this.

    The default value uses "$" which won't return nodes in the scene tree before "_ready()" is called. Use the "@onready" annotation to solve this.

    The default value uses "%" which won't return nodes in the scene tree before "_ready()" is called. Use the "@onready" annotation to solve this.

The default warning level for this warning is **Error**.
To modify it, set :ref:`Project Settings > Debug > GDScript > Warnings > Get Node Default Without Onready<class_ProjectSettings_property_debug/gdscript/warnings/get_node_default_without_onready>`.

When this warning occurs
------------------------

This warning may appear when attempting to assign the default value of a property to a node in the scene tree, like so:

.. code-block:: gdscript

    extends Area2D

>     # Will give warning GET_NODE_DEFAULT_WITHOUT_ONREADY.

Reviewer: I don't think "give warning" is an established terminology? Where has this been used elsewhere?

Author: From what I can see in the Godot Docs, there's not generally a lot of discussion of GDScript warnings. The verb "give" is used in at least one spot, [Static typing in GDScript](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/static_typing.html#warning-system):

> Godot gives you warnings about your code as you write it.

[Later on the same page](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/static_typing.html#unsafe-property-access-and-unsafe-method-access-warnings), "produce" is used:

> ```gdscript
> if "some_property" in node_2d:
> 	node_2d.some_property = 20  # Produces UNSAFE_PROPERTY_ACCESS warning.
> 
> if node_2d.has_method("some_function"):
> 	node_2d.some_function()  # Produces UNSAFE_METHOD_ACCESS warning.
> ```
> However, this code will produce `UNSAFE_PROPERTY_ACCESS` and `UNSAFE_METHOD_ACCESS` warnings...

In the dedicated page for the [GDScript warning system](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/warning_system.html#gdscript-warning-system), "trigger" is used:

> Godot will add an annotation above the corresponding line and the code won't trigger the corresponding warning anymore

---

So, overall it appears that "give", "produce", and "trigger" all have precedent in the docs currently.

Reviewer: Produce seems to be the most common, but I think we can leave it as is

Author: Alright, I will leave it for now then 🙂

    var my_collision_shape = $CollisionShape2D

> If you want a property to refer to another node in the scene tree, you may be inclined to reference it using :ref:`get_node() <class_Node_method_get_node>` (or the ``$`` and ``%`` shorthand versions) when setting its default value.

Reviewer: ```suggestion
If you want a property to refer to another node in the scene tree, you may be inclined to reference it using
:ref:`get_node() <class_Node_method_get_node>` (or the ``$`` and ``%`` shorthand versions) when setting its default value.
```
For all, keep to a reasonable length

GET_NODE_DEFAULT_WITHOUT_ONREADY
================================

This warning appears when a node's property is initialized to a node in the scene tree
without using the ``@onready`` annotation.

Depending on how you attempt to access the scene tree, the warning message will
be one of the following:

.. code-block:: none

    The default value uses "get_node()" which won't return nodes in the scene tree before "_ready()" is called. Use the "@onready" annotation to solve this.

    The default value uses "$" which won't return nodes in the scene tree before "_ready()" is called. Use the "@onready" annotation to solve this.

    The default value uses "%" which won't return nodes in the scene tree before "_ready()" is called. Use the "@onready" annotation to solve this.

The default warning level for this warning is **Error**.

> To modify it, set :ref:`Project Settings > Debug > GDScript > Warnings > Get Node Default Without Onready<class_ProjectSettings_property_debug/gdscript/warnings/get_node_default_without_onready>`.

Reviewer 2: I don't quite understand who this section is meant for. I thought this page was aimed at beginners to give them a short explanation on the common warnings.

I don't see how the default value is relevant to them.

Author: Eventually, the hope is to have every warning documented with examples, explanations, and fixes. This initial PR just adds pages for a few warnings that might be the most common, so that the review burden isn't as great.

Can you clarify what you mean by "I don't see how the default value is relevant to [beginners for a short explanation on the common warnings]"? I'm not sure I currently follow; the `GET_NODE_DEFAULT_WITHOUT_ONREADY` warning involves setting a property's default value with `get_node()` (or a shorthand equivalent) which is why default values are discussed here.

Reviewer 2: > Can you clarify what you mean by "I don't see how the default value is relevant to [beginners for a short explanation on the common warnings]"?

That was referring to this part:
>> The default warning level for this warning is **Error**.

The default value is already in the class reference, and there it's actually always up-to-date (although displaying it as int is a bit terrible I guess).

I don't think including this information here makes that much sense.

> Eventually, the hope is to have every warning documented with examples, explanations, and fixes.

Yeah I get that. I was talking about the general structure that you used here. I feel like we might just want to strike this initial section, and start with "When this warning occurs" directly.

>> This warning appears when a node's property is initialized to a node in the scene tree
without using the ``@onready`` annotation.

It seems like this is basically saying the same as the explanation you provide below.

>> Depending on how you attempt to access the scene tree, the warning message will
be one of the following:

I see no point in spelling out all variations. That's bound to get outdated, and users already know this message if they looked up the error.
(If you did that for SEO, that might be a point I guess. But we have well defined warning names, so that seems unnecessary to me 🤔.)

Author: > That was referring to this part

Ah, I see, thanks for the clarification! Thinking about it now, I agree that it might make more sense to just link to the class reference from these pages, and only specify the default value there. That way, updates to the default value won't make these pages outdated or inaccurate.

I recall there being discussion of adding links from these pages to the settings class reference; perhaps the inclusion of the default values was to make them flow more naturally? I'll see what I can find in the previous conversations, and if I can't find anything or what I find isn't especially compelling, I'll remove the default values and only link to the class reference.

---

> I feel like we might just want to strike this initial section, and start with "When this warning occurs" directly.
> ...
> It seems like this is basically saying the same as the explanation you provide below.

I definitely see where you're coming from here; the first sentence of the page is largely retreading ("pre-treading"?) the description given in the "When this warning occurs" section. The intent here is to provide a short (one- or two-sentence) description of the warning, along with the warning messages that might be displayed to the user (and thus that they might paste into a search engine to figure it out). 

The "short description" being largely repeated below has precedent in the Godot documentation. For example, in the [Node3D class reference](https://docs.godotengine.org/en/stable/classes/class_node.html), the initial short description says
> Base object in 3D space, inherited by all 3D nodes.

and just below, the long description starts with
> The **Node3D** node is the base representation of a node in 3D space. All other 3D nodes inherit from this class.

Bringing it back to the warning page, I suppose we could remove the short description of the warning and put the examples of warning messages in a section named something like "What this warning looks like". But given that other areas of the manual/class reference often start with an overview/intro section that is expanded upon in the following sections, I think it might be best to keep as-is.

---

> If you did that for SEO, that might be a point I guess

Yep, that was pretty much the reason. I would expect users to copy and paste the warning messages they receive into a search engine to look for help with it, so I want to make sure the messages themselves (not just the codes/codenames, like `GET_NODE_DEFAULT_WITHOUT_ONREADY`) cause these pages to show up in search results.

When this warning occurs
------------------------

This warning may appear when attempting to assign the default value of a property
to a node in the scene tree, like so:

.. code-block:: gdscript

    extends Area2D

    # Will give warning GET_NODE_DEFAULT_WITHOUT_ONREADY.
    var my_collision_shape = $CollisionShape2D

If you want a property to refer to another node in the scene tree, you may be inclined
to reference it using :ref:`get_node() <class_Node_method_get_node>`
(or the ``$`` and ``%`` shorthand versions) when setting its default value.

> However, properties' default values are evaluated and assigned *before* the scene tree is set up.

Reviewer 2: > *before* the scene tree is set up

That's not really the case though. If you instantiate a scene while the game is running, the scene tree is set up. It might be more accurate to say "before a node gets added to the scene tree"

This means that when Godot tries to assign that value, the node you're referring to may not be in

> the scene tree yet, and the property will be set to ``null`` instead.

Reviewer 2: The issue isn't the node you are referring to, it's the node with the script attached that is not in the scene tree.

GET_NODE_DEFAULT_WITHOUT_ONREADY
====================================

> This warning appears when the default value of a node's property is set to a location in the scene tree without using the ``@onready`` annotation.

Reviewer 3: This reads as... well, really overly technical to me. It took me a couple re-reads to figure out what it was trying to tell me.

Maybe:

```suggestion
This warning appears when a node's property is initialized to a location in the scene tree without using the ``@onready`` annotation.
```

Author: I like the suggestion for the sentence and will add that in, thanks!

As for the "property" vs "member variable" situation, I think I was mostly inspired by the class reference, where they are referred to as "properties". Granted, those are for built-in classes that are more of a black box for users, whereas the things we're talking about here users have direct access to modify and inspect. But for now, "property" is more succinct and used frequently in documentation, so I think it makes sense to keep it that way.

Depending on how you attempt to access the scene tree, the warning message will be one of the following:

.. code-block:: none

    The default value uses "$" which won't return nodes in the scene tree before "_ready()" is called. Use the "@onready" annotation to solve this.

    The default value uses "%" which won't return nodes in the scene tree before "_ready()" is called. Use the "@onready" annotation to solve this.

    The default value uses "get_node()" which won't return nodes in the scene tree before "_ready()" is called. Use the "@onready" annotation to solve this.

The default warning level for this warning is **Error**.
To modify it, set :ref:`Project Settings > Debug > GDScript > Warnings > Get Node Default Without Onready<class_ProjectSettings_property_debug/gdscript/warnings/get_node_default_without_onready>`.

When this warning occurs
------------------------

Instance properties can be set by declaring them outside of a method. Additionally, they can be given a default value using ``=``:

.. code-block:: gdscript

    extends Area2D

    var my_num = 3

The property ``my_num`` will always start out with a value of ``3`` in each instance of this class.

> GDScript also has methods to retrieve specific nodes from the scene tree: namely the :ref:`get_node() <class_Node_method_get_node>` method, and its shorthand versions ``$`` (and ``%`` for unique nodes). Thus, if you want to have a property's default value be a particular child node, it may be tempting to write something like the following:

Author: That's a good point. Originally, my thought was that the target audience for these docs is new programmers who could be confused by the warning messages themselves. They may be following tutorials and pasting code without fully understanding it, so having the full thought process leading to the code that causes the warning explained could help them.

That being said, explaining the basics of member variable initialization like this might be a bit *too* much background. This could be simplified to just the `get_node()` (and shorthand) methods, as well as how member variable initialization happens before building the scene tree. If there's a guide elsewhere in the docs that explains member variables for newcomers, we could link it here.

.. code-block::

    extends Area2D

    var my_collision_shape = $CollisionShape2D

> However, class instance variables' default values are evaluated and assigned before the scene tree is set up. This means that at the time of assigning the default value, the node may not be in the scene tree, and the variable will be set to ``null`` instead.

Author: I agree; since GDScript is so object-oriented, the fact that the variable belongs to a class is implied. I could see a few ways to tighten up the phrasing:

- "However, variables' default values are evaluated..."
- "However, a variable's default values are evaluated..."
- "However, the default values of a variable are evaluated..."

I'm not sure if removing "class instance" or "instance" could cause people to be confused about the behavior of local variables, though. Would it be proper to use the term "property" here instead of "variable"?

> However, class instance variables' default values are evaluated and assigned before the scene tree is set up. This means that at the time of assigning the default value, the node may not be in the scene tree, and the variable will be set to ``null`` instead.

Reviewer 4: I personally use "property" all the time when referring to member variables (so not local). It is much less ambiguous. The term is also fairly consistent in the class reference at least. So yes
